cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -448,20 +449,21 @@ public void upgradeApplications(FullHttpRequest request, HttpResponder responder
     // TODO: (CDAP-16910) Improve batch API performance as each application upgrade is an event independent of each
     //  other.
 
-    List<ApplicationId> appIds = decodeAndValidateBatchApplication(validateNamespace(namespaceId), request);
+    List<ApplicationId> appIds = decodeAndValidateBatchApplicationRecord(validateNamespace(namespaceId), request);
     Set<ArtifactScope> allowedArtifactScopes = getArtifactScopes(artifactScopes);
     List<ApplicationUpdateDetail> details = new ArrayList<>();
     for (ApplicationId appId : appIds) {
       ApplicationUpdateDetail updateDetail;
       try {
-        applicationLifecycleService.upgradeApplication(appId, createProgramTerminator(), allowedArtifactScopes,
-                                                       allowSnapshot);
-        updateDetail = new ApplicationUpdateDetail(appId, "upgrade successful.");
+        applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
+        updateDetail = new ApplicationUpdateDetail(appId);
       } catch (UnsupportedOperationException e) {
         String errorMessage = String.format("Application %s does not support upgrade.", appId);
-        updateDetail = new ApplicationUpdateDetail(appId, errorMessage);
-      } catch (InvalidArtifactException | NotFoundException e) {
-        updateDetail = new ApplicationUpdateDetail(appId, e.getMessage());
+        updateDetail = new ApplicationUpdateDetail(appId, new NotImplementedException(errorMessage));
+      } catch (NotFoundException e) {
+        updateDetail = new ApplicationUpdateDetail(appId, new NotFoundException(appId));
```

Comment:
Can you just pass in `e`?

Reply:
aah.. I was trying something else forgot to change this back. Thanks for noticing.